docs/guide/req-config.md
# 请求配置

这些是创建请求时可以用的配置选项。

只有 `url` 是必需的，如果没有指定 `method` 且没有指定 `adapter`，请求将默认使用 `GET` 方法。

```ts
const config = {
  // `url` 是用于请求的服务器 URL
  url: '/user',

  // `baseUrl` 将自动加在 `url` 前面，除非 `url` 是一个绝对 URL
  // 设置一个 `baseUrl` 便于为 un 实例的方法传递相对 URL
  baseUrl: 'https://some-domain.com/api/',

  // 自定义请求头
  headers: {
    'content-type': 'application/json',
  },

  // `params` 是与请求一起发送的 URL 参数
  // 必须是一个普通对象或一个 URLSearchParams 对象
  // 要使用 URLSearchParams 对象，请使用 core-js 提供的 polyfill
  // 可参考构建部分的说明或该仓库提供的 playground
  params: {
    ID: '12345',
  },

  // `paramsSerializer` 是可选方法，主要用于序列化 `params`
  // 默认使用 [query-string](https://github.com/sindresorhus/query-string) 序列化
  paramsSerializer: (params) => {
    /* 返回一个字符串 */
  },

  // `timeout` 指定请求超时的毫秒数
  // 如果请求时间超过 `timeout` 的值，则请求会被中断
  timeout: 1000, // 默认值是实际调用的 API 的默认值

  // `adapter` 允许自定义处理请求
  // 可以指定为 'request'、'upload' 和 'download' 三者之一
  // 也可以指定为一个方法，返回一个 Promise 并提供一个有效的响应
  // 如果你正在使用 un.request、un.download、un.upload、un.get 等别名方法
  // 无需再指定该键的值
  // adapter: 'request', // 默认值
  adapter: (config) => {
    /* ... */
  },

  // `validateStatus` 定义了对于给定的 HTTP 状态码是 resolve 还是 reject
  // 如果 `validateStatus` 返回 `true`、`null` 或 `undefined`
  // 则 promise 将会被 resolve，否则会被 reject
  validateStatus: function (status) {
    return status >= 200 && status < 300; // 默认值
  },

  // 用于取消请求
  signal: new AbortController().signal,

  // 用于取消请求
  cancelToken: new CancelToken(function (cancel) {
    /* ... */
  }),

  // 监听 HTTP Response Header 事件
  // 会比请求完成事件更早
  onHeadersReceived: (result) => {
    /* ... */
  },

  // request 使用
  // 创建请求时使用的方法
  method: 'GET', // 默认值

  // request 使用
  // `data` 是作为请求体被发送的数据
  // 必须是以下类型之一：string、object、ArrayBuffer、ArrayBufferView、URLSearchParams
  data: {
    firstName: 'Fred',
  },
  data: 'Country=Brasil&City=Belo Horizonte',

  // request 使用
  // 返回的数据类型
  // 如果设置为 json，会尝试对返回的数据做一次 JSON.parse
  dataType: 'json', // 默认值

  // request 使用
  // 响应的数据类型，选项包括 'text' 和 'arraybuffer'
  responseType: 'text', // 默认值

  // request 使用
  // 是否开启 http2
  enableHttp2: false, // 默认值

  // request 使用
  // 是否开启 quic
  enableQuic: false, // 默认值

  // request 使用
  // 是否开启缓存
  enableCache: false, // 默认值

  // request 使用
  // 是否开启 HttpDNS 服务
  enableHttpDNS: false, // 默认值

  // request 使用
  // HttpDNS 服务商 Id
  httpDNSServiceId: '',

  // request 使用
  // 是否开启 transfer-encoding chunked
  enableChunked: false, // 默认值

  // request 使用
  // 是否在 wifi 下使用移动网络发送请求
  forceCellularNetwork: false, // 默认值

  // request 使用
  // 是否验证 ssl 证书
  sslVerify: true, // 默认值

  // request 使用
  // 跨域请求时是否需要使用凭证
  withCredentials: false, // 默认值

  // request 使用
  // 是否在 DNS 解析时优先使用 ipv4
  firstIpv4: false, // 默认值

  // request 使用
  // 监听 Transfer-Encoding Chunk Received 事件
  // 当接收到新的 chunk 时触发
  onChunkReceived: (response) => {
    /* ... */
  },

  // upload 使用
  // 需要上传的文件列表，files 和 filePath 必填一个
  // 使用该参数时，filePath 和 name 无效
  // 不支持小程序
  files: [],

  // upload 使用
  // 文件类型
  fileType: 'image', // image, video, audio

  // upload 使用
  // 文件对象
  file: new File(),

  // upload 使用
  // 文件路径，files 和 filePath 必填一个
  //
  // download 使用
  // 文件下载后存储的本地路径
  filePath: '/fake/path',

  // upload 使用
  // 文件名称
  name: 'fake-file.png',

  // upload 使用
  // 一个对象，会作为 HTTP 请求中其它额外的 form data
  formData: Record<string, any>,

  // download 使用
  // 下载进度变化时触发
  // 优先级 onDownloadProgress > onDownloadProgressUpdate > onProgress > onProgressUpdate
  onDownloadProgress: UnOnProgress,

  // download 使用
  // 下载进度变化时触发
  // 优先级 onDownloadProgress > onDownloadProgressUpdate > onProgress > onProgressUpdate
  onDownloadProgressUpdate: UnOnProgress,

  // upload 使用
  // 上传进度变化时触发
  // 优先级 onUploadProgress > onUploadProgressUpdate > onProgress > onProgressUpdate
  onUploadProgress: UnOnProgress,

  // upload 使用
  // 上传进度变化时触发
  // 优先级 onUploadProgress > onUploadProgressUpdate > onProgress > onProgressUpdate
  onUploadProgressUpdate: UnOnProgress,

  // upload / download 使用
  // 上传/下载进度变化时触发
  // 优先级 onUploadProgress / onDownloadProgress > onUploadProgressUpdate / onDownloadProgressUpdate > onProgress > onProgressUpdate
  onProgress: UnOnProgress,

  // upload / download 使用
  // 上传/下载进度变化时触发
  // 优先级 onUploadProgress / onDownloadProgress > onUploadProgressUpdate / onDownloadProgressUpdate > onProgress > onProgressUpdate
  onProgressUpdate: UnOnProgress,
};
```


docs/api/request-config.md
# 请求配置

以下是创建请求时可以用的配置选项。只有 `url` 是必需的。如果没有指定 `method` 且没有指定 `adapter`，请求将默认使用 `GET` 方法。

```typescript
{
  // `url` 是用于请求的服务器 URL
  url: '/user',

  // `baseUrl` 将自动加在 `url` 前面，除非 `url` 是一个绝对 URL
  // 设置一个 `baseUrl` 便于为实例方法传递相对 URL
  baseUrl: 'https://some-domain.com/api/',

  // 自定义请求头
  // 不能设置 Referer
  headers: {
    'content-type': 'application/json',
  },

  // `params` 是与请求一起发送的 URL 参数
  // 必须是一个普通对象或一个 URLSearchParams 对象
  // 要使用 URLSearchParams 对象，请使用 core-js 提供的 polyfill
  // 可参考构建与环境支持部分的说明或该仓库提供的 playground
  params: {
    ID: '12345',
  },

  // `paramsSerializer` 是可选方法，主要用于序列化 `params`
  // 默认使用 [fast-querystring](https://github.com/anonrig/fast-querystring) 序列化
  // [qs](https://github.com/ljharb/qs) v6.10.0 开始引入了 `get-intrinsic`，结合微信小程序和微信小程序插件使用时会出现报错，如有需要可以使用 v6.9.7
  // [query-string](https://github.com/sindresorhus/query-string) v8.1.0 使用了支付宝小程序不支持的语法，如无支付宝小程序需求也可以使用 query-string，它体积比 qs 小，性能比 qs 好
  paramsSerializer: (params) => {
    /* 返回一个字符串 */
  },

  // `timeout` 指定请求超时的毫秒数
  // 如果请求时间超过 `timeout` 的值，则请求会被中断
  // 要设置永不超时，可以将其设置为 Number.POSITIVE_INFINITY
  // 默认值是实际调用的 API 的默认值，见 https://uniapp.dcloud.net.cn/collocation/manifest.html#networktimeout
  timeout: 1000,

  // `adapter` 允许自定义处理请求
  // 可以指定为 'request'、'upload' 和 'download' 三者之一
  // 也可以指定为一个方法，返回一个 Promise 并提供一个有效的响应
  // 如果你正在使用 un.request、un.download、un.upload、un.get 等别名方法，则无需再指定该键的值
  // 默认值是 'request'
  adapter: (config) => { /* ... */ },

  // `validateStatus` 定义了对于给定的 HTTP 状态码该 resolve 还是 reject
  // 如果 `validateStatus` 返回 `true`、`null` 或 `undefined`
  // 则 promise 将会被 resolve，否则会被 reject
  validateStatus: function (status) {
    return status >= 200 && status < 300; // 默认值
  },

  // 用于取消请求
  // 可参考取消请求部分的说明
  signal: new AbortController().signal,

  // 用于取消请求
  // 可参考取消请求部分的说明
  cancelToken: new CancelToken(function (cancel) {
    /* ... */
  }),

  // 监听 HTTP Response Header 事件
  // 会比请求完成事件更早
  onHeadersReceived: (result) => {
    /* ... */
  },

  // request 使用
  // 创建请求时使用的方法
  // 默认值是 'GET'
  method: 'GET',

  // request 使用
  // `data` 是作为请求体被发送的数据
  // 必须是以下类型之一：string、object、ArrayBuffer、ArrayBufferView、URLSearchParams
  data: {
    firstName: 'Fred',
  },
  // 这也是可行的
  // data: 'Country=Brasil&City=Belo Horizonte',

  // request 使用
  // 返回的数据类型
  // 如果设置为 json，会尝试对返回的数据做一次 JSON.parse
  // 默认值是 'json'
  dataType: 'json',

  // request 使用
  // 响应的数据类型，选项包括 'text' 和 'arraybuffer'
  // 默认值是 'text'
  responseType: 'text',

  // request 使用
  // 是否开启 http2
  // 默认值是 false
  enableHttp2: false,

  // request 使用
  // 是否开启 quic
  // 默认值是 false
  enableQuic: false,

  // request 使用
  // 是否开启缓存
  // 默认值是 false
  enableCache: false,

  // request 使用
  // 是否开启 HttpDNS 服务
  // 默认值是 false
  enableHttpDNS: false,

  // request 使用
  // HttpDNS 服务商 Id
  httpDNSServiceId: '',

  // request 使用
  // 是否开启 transfer-encoding chunked
  // 默认值是 false
  enableChunked: false,

  // request 使用
  // 是否在 wifi 下使用移动网络发送请求
  // 默认值是 false
  forceCellularNetwork: false,

  // request 使用
  // 是否验证 ssl 证书
  // 默认值是 true
  sslVerify: true,

  // request 使用
  // 跨域请求时是否需要使用凭证
  // 默认值是 false
  withCredentials: false,

  // request 使用
  // 是否在 DNS 解析时优先使用 ipv4
  // 默认值是 false
  firstIpv4: false,

  // request 使用
  // 监听 Transfer-Encoding Chunk Received 事件
  // 当接收到新的 chunk 时触发
  onChunkReceived: (response) => { /* ... */ },

  // upload 使用
  // 需要上传的文件列表，files 和 filePath 必填一个
  // 使用该参数时，filePath 和 name 无效
  // 不支持小程序
  files: [],

  // upload 使用
  // 文件类型
  fileType: 'image', // image, video, audio

  // upload 使用
  // 文件对象
  file: new File(),

  // upload 使用
  // 文件路径，files 和 filePath 必填一个
  //
  // download 使用
  // 文件下载后存储的本地路径
  filePath: '/fake/path',

  // upload 使用
  // 文件名称
  name: 'fake-file.png',

  // upload 使用
  // 一个对象，会作为 HTTP 请求中其它额外的 form data
  formData: Record<string, any>,

  // download 使用
  // 下载进度变化时触发
  // 优先级 onDownloadProgress > onDownloadProgressUpdate > onProgress > onProgressUpdate
  onDownloadProgress: UnOnProgress,

  // download 使用
  // 下载进度变化时触发
  // 优先级 onDownloadProgress > onDownloadProgressUpdate > onProgress > onProgressUpdate
  onDownloadProgressUpdate: UnOnProgress,

  // upload 使用
  // 上传进度变化时触发
  // 优先级 onUploadProgress > onUploadProgressUpdate > onProgress > onProgressUpdate
  onUploadProgress: UnOnProgress,

  // upload 使用
  // 上传进度变化时触发
  // 优先级 onUploadProgress > onUploadProgressUpdate > onProgress > onProgressUpdate
  onUploadProgressUpdate: UnOnProgress,

  // upload / download 使用
  // 上传/下载进度变化时触发
  // 优先级 onUploadProgress / onDownloadProgress > onUploadProgressUpdate / onDownloadProgressUpdate > onProgress > onProgressUpdate
  onProgress: UnOnProgress,

  // upload / download 使用
  // 上传/下载进度变化时触发
  // 优先级 onUploadProgress / onDownloadProgress > onUploadProgressUpdate / onDownloadProgressUpdate > onProgress > onProgressUpdate
  onProgressUpdate: UnOnProgress,
};
```


CHANGELOG.md
# 改动日志

## 0.18.0 (2023-01-03)

- feat!: 默认使用 [fast-querystring](https://github.com/anonrig/fast-querystring) 而不是 [query-string](https://github.com/sindresorhus/query-string) 序列化 `params`
  - `query-string@8.1.0` 和它所依赖的库存在 `try {} catch {}` 的用法，该用法不受支付宝小程序支持

## 0.17.0 (2023-11-04)

- feat!: 现在要求 `node>=18`

## 0.16.1 (2023-08-08)

- fix: 移除默认的 timeout 值，修复 H5 端请求立即超时的问题，关闭 [#37](https://github.com/uni-helper/uni-network/issues/37)

## 0.16.0 (2023-07-03)

- feat!: 默认使用 [query-string](https://github.com/sindresorhus/query-string) 而不是 [qs](https://github.com/ljharb/qs) 序列化 `params`

  - `query-string@8.1.0` 支持 `node >= 14.16`，没有过多的历史包袱，而 `qs@6.11.2` 至今还在支持 `node >= 0.6`
  - `qs@6.10.0` 开始引入了 `get-intrinsic`，结合微信小程序和微信小程序插件使用时会出现报错，参考 [#31](https://github.com/uni-helper/uni-network/issues/31)，而 `query-string@8.1.0` 没有这个问题
  - 如果你的 `params` 对象内某个键的值为对象、数组或 Date，表现行为会不一致
  - 如果你更倾向于使用该库 0.15.0 版本或更早版本的默认设置，请安装 `qs@6.9.7`，并设置 `paramsSerializer`

  ```typescript
  {
    ...,
    paramsSerializer: (params: Record<string, any>) => {
      return Object.prototype.toString.call(params).includes('URLSearchParams')
        ? params.toString()
        : qs.stringify(params);
    },
    ...,
  }
  ```

## 0.15.0 (2023-04-16)

- feat(composables)!: 调整返回的 `error` Ref 类型为 `unknown`，对齐 `vueuse@10` 行为，查看 [vueuse/vueuse#2807](https://github.com/vueuse/vueuse/pull/2807) 了解更多
- feat(composables)!: 使用 `PromiseLike` 替换 `Promise`，对齐 `vueuse@10` 行为，查看 [vueuse/vueuse#2485](https://github.com/vueuse/vueuse/pull/2485) 了解更多
- feat(composables): 增加 `onFinish` 选项，对齐 `vueuse@10` 行为，查看 [vueuse/vueuse#2829](https://github.com/vueuse/vueuse/pull/2829) 了解更多
- feat(composables): 增加 `initialData` 和 `resetOnExecute` 选项，对齐 `vueuse@10` 行为，查看 [vueuse/vueuse#2791](https://github.com/vueuse/vueuse/pull/2791) 了解更多
- fix(composables): 修复 `UnConfig` 类型使用

## 0.14.0 (2023-02-28)

- feat: 新增组合式函数

## 0.13.3 (2023-02-08)

和 0.13.1 一样，但又修复了一个 ci 问题。

## 0.13.2 (2023-02-08)

和 0.13.1 一样，但修复了一个 ci 问题。

## 0.13.1 (2023-02-08)

- fix: 修复了 `un.download` 和 `un.upload` 没有正确设置 `adapter` 的问题，感谢 [@edazh](https://github.com/edazh) 在 [#25](https://github.com/uni-helper/uni-network/pull/25) 的贡献
- build: 构建工具切换到 `unbuild`

## 0.13.0 (2023-02-06)

调整了部分代码风格，同时更新了文档。

## 0.12.6 (2023-01-10)

- fix: 处理 `statuses` 可能抛出的错误，感谢 [@edazh](https://github.com/edazh) 在 [#18](https://github.com/uni-helper/uni-network/pull/18) 的贡献

## 0.12.5 (2023-01-04)

- build: 使用 `rollup` 构建
- perf: `mergeConfig` 实现调整

## 0.12.4 (2022-12-30)

- fix: 移除构建中 `qs` 相关代码以修复构建

## 0.12.3 (2022-12-30)

误发布，请勿使用。

## 0.12.2 (2022-12-30)

- fix: 修复构建

## 0.12.1 (2022-12-29)

- perf: 明确导入类型
- perf: 分离 `HttpStatusCode`
- perf: 暴露 `mergeConfig` 和 `HttpStatusCode`

## 0.12.0 (2022-12-28)

- feat!: 要求 `node >= 14.18`，这是为了对标 `rollup` 和 `vite`
- feat: esm 优先，但仍然提供 cjs 支持
- fix: 修复导出
- perf: 构建包含 `ramda`、`statuses` 和 `qs` 相关代码

## 0.11.1 (2022-11-16)

- fix: 修复构建

## 0.11.0 (2022-11-16)

- feat!: 迁移到 `@uni-helper/uni-network`，`Uan` 前缀调整为 `Un`
- perf: 移除 `lodash-es`

## 0.10.2 (2022-10-18)

- perf: 优化 `params` 类型
- perf: 分离 `mergeConfig` 方法
- fix: 修复 `UanCanceler` 类型

## 0.10.1 (2022-10-12)

- fix: 修复构建

## 0.10.0 (2022-10-11)

- feat!: 重命名一些 `request` 为 `task` 避免误导
- feat!: 调整类型，移除了 `UanBaseXxx`、`UanRequestXxx`、`UanDownloadXxx`、`UanUploadXxx` 等类型，可直接使用 `UanXxx`
- feat: 支持 `onProgress`、`onProgressUpdate`、`onDownloadProgress`、`onUploadProgressUpdate`、`onUploadProgress`、`onUploadProgressUpdate`

## 0.8.0 (2022-10-07)

- feat: 增加枚举数据
- fix: 修复类型定义
- perf: 优化类型定义
- feat!: 导出的 `isCancel` 调整为 `isUanCancel`

## 0.7.1 (2022-09-30)

- fix: 修复了构建不正常的问题

## 0.7.0 (2022-09-29)

- feat!: 现在要求使用 `node >= 14.16`
- feat!: 现在构建目标是 `esnext`
- fix: 修复了构建不正常的问题

## 0.6.0

- feat!: 重命名为 `uni-app-network`

## 0.4.8

- perf: 更新 `utils/extend`，匹配 `axios` 改动

## 0.4.7

- fix: 修复导出

## 0.4.6

- fix: 修复导出

## 0.4.5

- fix: 修复类型

## 0.4.4

- fix: 修复类型

## 0.4.3

- fix: 导出全部类型

## 0.4.2

修复说明文档错误。

## 0.4.1

- fix: 修复类型

## 0.4.0

- feat: 迁移到 TypeScript
- feat: 增加导出
- perf: 移除多余的向后兼容代码
- fix: 修复构建
- test: 修复测试

## 0.3.2

- fix: 修复类型

## 0.3.1

- fix: 修复方法没有正确挂载的问题
- fix: 修复错误解构的问题

## 0.3.0

- rewrite: 完全重写

## 0.2.1

- fix: 修复类型定义

## 0.2.0

- feat: `request` 支持使用字符串作为第一个参数
- feat: `useRequest` 支持使用字符串作为第一个参数
- fix: 修复类型定义
- fix: 修复导出

## 0.1.1

- fix: 修复 `request` 配置错误的问题

## 0.1.0

- feat: 提供 `request`
- feat: 提供 `useRequest`
- feat: 提供 `requestAdapter`


src/types/config.ts
import { UnCancelToken } from '../core/UnCancelToken';
import type {
  UnData,
  UnHeaders,
  UnMethod,
  UnDataType,
  UnResponseType,
  UnGenericAbortSignal,
  UnOnProgress,
  UnFileType,
  UnParamsSerializer,
  UnValidateStatus,
  UnFile,
  UnParams,
} from './common';
import type { UnAdapter } from './adapter';

export interface UnConfig<T = UnData, D = UnData> {
  /** 用于请求的服务器 URL */
  url?: string;
  /**
   * 创建请求时使用的方法
   *
   * 默认为 'GET'
   */
  method?: UnMethod;
  /** 自动加在 `url` 前面，除非 `url` 是一个绝对 URL */
  baseUrl?: string;
  /** 自定义请求头，不能设置 Referer */
  headers?: UnHeaders;
  /** 与请求一起发送的 URL 参数 */
  params?: UnParams;
  /**
   * 可选方法，主要用于序列化 `params`
   *
   * 默认使用 [fast-querystring](https://github.com/anonrig/fast-querystring) 序列化
   *
   * [qs](https://github.com/ljharb/qs) v6.10.0 开始引入了
   * `get-intrinsic`，结合微信小程序和微信小程序插件使用时会出现报错，如有需要可以使用 v6.9.7
   *
   * [query-string](https://github.com/sindresorhus/query-string) v8.1.0
   * 使用了支付宝小程序不支持的语法，如无支付宝小程序需求也可以使用 query-string，它体积比 qs 小，性能比 qs 好
   */
  paramsSerializer?: UnParamsSerializer;
  /**
   * 作为请求体被发送的数据
   *
   * 必须是以下类型之一：string、ArrayBuffer、Record<string, any>
   */
  data?: D;
  /**
   * 指定请求超时的毫秒数
   *
   * 如果请求时间超过 `timeout` 的值，则请求会被中断
   *
   * 要设置永不超时，可以将其设置为 Number.POSITIVE_INFINITY
   *
   * 默认值是实际调用的 API 的默认值
   */
  timeout?: number;
  /**
   * 允许自定义处理请求
   *
   * 可以指定为 'request'、`upload` 和 `download` 三者之一
   *
   * 也可以指定为一个方法，返回一个 Promise 并提供一个有效的响应
   *
   * 如果你正在使用 un.request、un.download、un.upload、un.get 等别名方法，则无需再指定该键的值
   *
   * 默认为 'request'
   */
  adapter?: 'request' | 'download' | 'upload' | UnAdapter<T, D>;
  /**
   * 定义了对于给定的 HTTP 状态码该 resolve 还是 reject
   *
   * 如果 `validateStatus` 返回 `true`、`null` 或 `undefined`
   *
   * 则 promise 将会被 resolve，否则会被 reject
   *
   * 默认为 (status) => status >= 200 && status < 300
   */
  validateStatus?: UnValidateStatus;
  /** 用于取消请求 */
  signal?: UnGenericAbortSignal;
  /** 用于取消请求 */
  cancelToken?: UnCancelToken<T, D>;
  /**
   * 监听 HTTP Response Header 事件
   *
   * 会比请求完成事件更早
   */
  onHeadersReceived?: (response?: { headers?: UnHeaders }) => void;

  /**
   * Request 使用
   *
   * 服务器返回数据的类型
   *
   * 如果设置为 json，会尝试对返回的数据做一次 JSON.parse
   *
   * 默认为 json
   */
  dataType?: UnDataType;
  /**
   * Request 使用
   *
   * 响应的数据类型
   *
   * 默认为 text
   */
  responseType?: UnResponseType;
  /**
   * Request 使用
   *
   * 是否开启 http2
   *
   * 默认为 false
   */
  enableHttp2?: boolean;
  /**
   * Request 使用
   *
   * 是否开启 quic
   *
   * 默认为 false
   */
  enableQuic?: boolean;
  /**
   * Request 使用
   *
   * 是否开启缓存
   *
   * 默认为 false
   */
  enableCache?: boolean;
  /**
   * Request 使用
   *
   * 是否开启 HttpDNS 服务
   *
   * 默认为 false
   */
  enableHttpDNS?: boolean;
  /**
   * Request 使用
   *
   * HttpDNS 服务商 Id
   */
  httpDNSServiceId?: string;
  /**
   * Request 使用
   *
   * 是否开启 transfer-encoding chunked
   *
   * 默认为 false
   */
  enableChunked?: boolean;
  /**
   * Request 使用
   *
   * 是否在 wifi 下使用移动网络发送请求
   *
   * 默认为 false
   */
  forceCellularNetwork?: boolean;
  /**
   * Request 使用
   *
   * 是否验证 ssl 证书
   *
   * 默认为 true
   */
  sslVerify?: boolean;
  /**
   * Request 使用
   *
   * 跨域请求时是否需要使用凭证
   *
   * 默认为 false
   */
  withCredentials?: boolean;
  /**
   * Request 使用
   *
   * 是否在 DNS 解析时优先使用 ipv4
   *
   * 默认为 false
   */
  firstIpv4?: boolean;
  /**
   * Request 使用
   *
   * 监听 Transfer-Encoding Chunk Received 事件
   *
   * 当接收到新的 chunk 时触发
   */
  onChunkReceived?: (response?: { data?: ArrayBuffer }) => void;

  /**
   * Upload 使用
   *
   * 需要上传的文件列表，files 和 filePath 必填一个
   *
   * 使用该参数时，filePath 和 name 无效
   *
   * 不支持小程序
   */
  files?: UnFile[];
  /**
   * Upload 使用
   *
   * 文件类型
   */
  fileType?: UnFileType;
  /**
   * Upload 使用
   *
   * 文件对象
   */
  file?: File;
  /**
   * Upload 使用
   *
   * 文件路径，files 和 filePath 必填一个
   *
   * Download 使用
   *
   * 文件下载后存储的本地路径
   */
  filePath?: string;
  /**
   * Upload 使用
   *
   * 文件名称
   */
  name?: string;
  /**
   * Upload 使用
   *
   * 一个对象，会作为 HTTP 请求中其它额外的 form data
   */
  formData?: Record<string, any>;
  /**
   * Download 使用
   *
   * 下载进度变化时触发
   *
   * 优先级 onDownloadProgress > onDownloadProgressUpdate > onProgress > onProgressUpdate
   */
  onDownloadProgress?: UnOnProgress;
  /**
   * Download 使用
   *
   * 下载进度变化时触发
   *
   * 优先级 onDownloadProgress > onDownloadProgressUpdate > onProgress > onProgressUpdate
   */
  onDownloadProgressUpdate?: UnOnProgress;
  /**
   * Upload 使用
   *
   * 上传进度变化时触发
   *
   * 优先级 onUploadProgress > onUploadProgressUpdate > onProgress > onProgressUpdate
   */
  onUploadProgress?: UnOnProgress;
  /**
   * Upload 使用
   *
   * 上传进度变化时触发
   *
   * 优先级 onUploadProgress > onUploadProgressUpdate > onProgress > onProgressUpdate
   */
  onUploadProgressUpdate?: UnOnProgress;
  /**
   * Upload / download 使用
   *
   * 上传/下载进度变化时触发
   *
   * 优先级 onUploadProgress / onDownloadProgress > onUploadProgressUpdate / onDownloadProgressUpdate >
   * onProgress > onProgressUpdate
   */
  onProgress?: UnOnProgress;
  /**
   * Upload / download 使用
   *
   * 上传/下载进度变化时触发
   *
   * 优先级 onUploadProgress / onDownloadProgress > onUploadProgressUpdate / onDownloadProgressUpdate >
   * onProgress > onProgressUpdate
   */
  onProgressUpdate?: UnOnProgress;
  [key: string]: any;
}
